File: Components/PlaneFinder.cs
using System;
using System.Collections.Generic;
using Grasshopper.Kernel;
using Rhino.Geometry;

namespace Enzyme.Components
{
    public class PlaneFinder : GH_Component
    {
        public PlaneFinder()
          : base("Surface Plane Finder", "plane_finder",
              "Finds a plane from a planar surface or brep face.",
              "Enzyme", "Pattern")
        {
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddGeometryParameter("x", "x", "Input geometry", GH_ParamAccess.list);
            pManager[0].Optional = true;
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddVectorParameter("plane_normal", "plane_normal", "Normal of the plane", GH_ParamAccess.item);
            pManager.AddVectorParameter("reverse_normal", "reverse_normal", "Reversed normal", GH_ParamAccess.item);
            pManager.AddPlaneParameter("output_plane", "output_plane", "Output plane", GH_ParamAccess.item);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            List<GeometryBase> xList = new List<GeometryBase>();
            if (!DA.GetDataList(0, xList)) return;

            string component_name = "Surface Plane Finder";
            string component_version = "v1.0";

            GeometryBase input_item = null;
            if (xList.Count > 0 && xList[0] != null)
            {
                input_item = xList[0];
            }

            Surface surface = null;

            if (input_item != null)
            {
                if (input_item is Surface srf)
                {
                    surface = srf;
                }
                else if (input_item is Brep brep)
                {
                    if (brep.Faces.Count == 1)
                    {
                        surface = brep.Faces[0].UnderlyingSurface();
                    }
                    else
                    {
                        this.Message = $"{component_name} {component_version}\nInput is a polysurface.";
                    }
                }
                else
                {
                    this.Message = $"{component_name} {component_version}\nInvalid input type.";
                }
            }
            else if (xList.Count == 0 || xList[0] == null)
            {
                this.Message = $"{component_name} {component_version}\nNo valid surface provided.";
            }

            if (surface != null)
            {
                double tol = Rhino.RhinoDoc.ActiveDoc != null ? Rhino.RhinoDoc.ActiveDoc.ModelAbsoluteTolerance : 0.001;
                
                if (surface.IsPlanar(tol))
                {
                    double u = surface.Domain(0).Mid;
                    double v = surface.Domain(1).Mid;

                    Vector3d normal = surface.NormalAt(u, v);
                    normal.Unitize();

                    Vector3d reverse_normal = -normal;
                    Point3d centroid;

                    if (surface is PlaneSurface planeSrf)
                    {
                        BoundingBox bbox = planeSrf.GetBoundingBox(true);
                        centroid = bbox.Center;
                    }
                    else
                    {
                        var brepFromSurface = Brep.CreateFromSurface(surface);
                        if (brepFromSurface != null)
                        {
                            var amp = AreaMassProperties.Compute(brepFromSurface);
                            if (amp != null)
                            {
                                centroid = amp.Centroid;
                            }
                            else
                            {
                                centroid = surface.PointAt(u, v);
                            }
                        }
                        else
                        {
                            centroid = surface.PointAt(u, v);
                        }
                    }

                    Plane plane = new Plane(centroid, normal);

                    DA.SetData(0, normal);
                    DA.SetData(1, reverse_normal);
                    DA.SetData(2, plane);

                    this.Message = $"{component_name} {component_version}\nSurface is planar.";
                }
                else
                {
                    this.Message = $"{component_name} {component_version}\nSurface is NOT planar.";
                }
            }
        }

        protected override System.Drawing.Bitmap Icon
        {
            get { return null; }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("f6e9bc6f-cd1e-4581-9b19-c0202ad9d968"); }
        }
    }
}

File: ScriptsToProcess/NEW BATCH TO PROCESS/plane_finder.py
import Rhino.Geometry as rg

# Manually set the component's name, nickname, and version
component_name = "Surface Plane Finder"
component_nickname = "plane_finder"
component_version = "v1.0"

# Check if the input `x` is a list
if isinstance(x, list):
    if len(x) > 0:
        input_item = x[0]  # Get the first item in the list
    else:
        input_item = None  # The list is empty
else:
    input_item = x  # Input is not a list

# Initialize `surface` as None
surface = None

# Check the type of the input and process accordingly
if input_item is not None:
    if isinstance(input_item, rg.Surface):
        # Input is a Surface
        surface = input_item
    elif isinstance(input_item, rg.Brep):
        # Input is a Brep, try to extract the first face
        if input_item.Faces.Count == 1:
            surface = input_item.Faces[0].UnderlyingSurface()  # Convert to Surface
        else:
            ghenv.Component.Message = f"{component_name} {component_version}\nInput is a polysurface."
    else:
        ghenv.Component.Message = f"{component_name} {component_version}\nInvalid input type."

# Proceed if a valid surface is extracted
if surface is not None:
    if surface.IsPlanar():
        # Get midpoints of the domain
        u = surface.Domain(0).Mid
        v = surface.Domain(1).Mid

        # Compute the normal
        normal = surface.NormalAt(u, v)
        normal.Unitize()

        # Reverse the normal
        reverse_normal = -normal

        # Compute the centroid
        if isinstance(surface, rg.PlaneSurface):
            # For PlaneSurface, use the bounding box to calculate the centroid
            bbox = surface.GetBoundingBox(True)
            centroid = bbox.Center
        else:
            # For other surfaces, use AreaCentroid
            success, centroid = surface.AreaCentroid()
            if not success:
                centroid = surface.PointAt(u, v)

        # Create a plane at the centroid with the normal
        plane = rg.Plane(centroid, normal)

        # Output
        plane_normal = normal
        reverse_normal = reverse_normal
        output_plane = plane
        ghenv.Component.Message = f"{component_name} {component_version}\nSurface is planar."
    else:
        plane_normal = None
        reverse_normal = None
        output_plane = None
        ghenv.Component.Message = f"{component_name} {component_version}\nSurface is NOT planar."
else:
    plane_normal = None
    reverse_normal = None
    output_plane = None
    ghenv.Component.Message = f"{component_name} {component_version}\nNo valid surface provided."

# Optional: Print debugging information
print(f"Component Name: {component_name}")
print(f"Component Nickname: {component_nickname}")
print(f"Component Version: {component_version}")
print(f"Input type: {type(input_item)}")

ghenv.Component.Name = component_name
ghenv.Component.NickName = component_nickname